src/Microsoft.AspNetCore.Routing/Template/TemplateMatcher.cs

```diff
@@ -82,20 +81,25 @@ public bool TryMatch(PathString path, RouteValueDictionary values)
             //
             // On the second pass, we're almost certainly going to match the URL, so go ahead and allocate the 'values'
             // and start capturing strings. 
-            foreach (var requestSegment in pathTokenizer)
+            foreach (var pathSegment in pathTokenizer)
             {
+                if (pathSegment.Length == 0)
+                {
+                    return false;
+                }
```

Comment:
Yep. It was a matter of not taking into account pathSegment.Length == 0 which can only happen at the start or in the middle of a request path if it contains any number of consecutive //. For trailing `/` the PathTokenizer will return false